zipline/utils/run_algo.py

```diff
@@ -0,0 +1,326 @@
+    """
+    if algotext is not None:
+        if local_namespace:
+            ip = get_ipython()  # noqa
+            namespace = ip.user_ns
+        else:
+            namespace = {}
+
+        for assign in defines:
+            try:
+                name, value = assign.split('=', 2)
+            except ValueError:
+                raise ValueError(
+                    'invalid define %r, should be of the form name=value' %
+                    assign,
+                )
+            try:
+                # evaluate in the same namespace so names may refer to
+                # eachother
+                namespace[name] = eval(value, namespace)
+            except Exception as e:
+                raise ValueError(
+                    'failed to execute definition for name %r: %s' % (name, e),
+                )
+    elif defines:
+        raise _RunAlgoError(
+            'cannot pass define without `algotext`',
+            "cannot pass '-D' / '--define' without '-t' / '--algotext'",
+        )
+    else:
+        namespace = {}
+        if algofile is not None:
+            algotext = algofile.read()
+
+    if print_algo:
+        if PYGMENTS:
+            highlight(
+                algotext,
+                PythonLexer(),
+                TerminalFormatter(),
+                outfile=sys.stdout,
+            )
+        else:
+            click.echo(algotext)
+
+    if bundle is not None:
+        bundle_data = load(
+            bundle,
+            environ,
+            bundle_timestamp,
+        )
+
+        prefix, connstr = re.split(
+            r'sqlite:///',
+            str(bundle_data.asset_finder.engine.url),
+            maxsplit=1,
+        )
+        if prefix:
+            raise ValueError(
+                "invalid url %r, must begin with 'sqlite:///'" %
+                str(bundle_data.asset_finder.engine.url),
+            )
+        env = TradingEnvironment(asset_db_path=connstr)
+        data = DataPortal(
+            env,
+            equity_minute_reader=bundle_data.minute_bar_reader,
+            equity_daily_reader=bundle_data.daily_bar_reader,
+            adjustment_reader=bundle_data.adjustment_reader,
+        )
+
+    perf = TradingAlgorithm(
+        namespace=namespace,
+        capital_base=capital_base,
+        start=start,
+        end=end,
+        env=env,
+        **{
+            'initialize': initialize,
+            'handle_data': handle_data,
+            'before_trading_start': before_trading_start,
+            'analyze': analyze,
+        } if algotext is None else {
+            'algo_filename': algofile,
+            'script': algotext,
+        }
+    ).run(
+        data,
+        overwrite_sim_params=False,
+    )
+
+    if output == '-':
+        click.echo(str(perf))
+    elif output != os.devnull:  # make the zipline magic not write any data
+        perf.to_pickle(output)
+
+    return perf
+
+
+def _load_extensions(default, extensions, strict, environ):
```

Comment:
yeah, I pulled this out so they could share. Looks like I missed a critical step ;)